Core/GDCore/Project/ObjectFolderOrObject.h

```diff
@@ -0,0 +1,197 @@
+
+  /**
+   * \brief Returns the object behind the instance.
+   */
+  gd::Object& GetObject() const { return *object; }
+
+  /**
+   * \brief Returns true if the instance represents a folder.
+   */
+  bool IsFolder() const { return !folderName.empty(); }
+  /**
+   * \brief Returns the name of the folder.
+   */
+  const gd::String& GetFolderName() const { return folderName; }
+
+  /**
+   * \brief Set the folder name. Does nothing if called on an instance not
+   * representing a folder.
+   */
+  void SetFolderName(const gd::String& name);
+
+  /**
+   * \brief Returns true if the instance represents the object with the given
+   * name or if any of the children does (recursive search).
+   */
+  bool HasObjectNamed(const gd::String& name);
+  /**
+   * \brief Returns the child instance holding the object with the given name
+   * (recursive search).
+   */
+  ObjectFolderOrObject& GetObjectNamed(const gd::String& name);
+
+  /**
+   * \brief Returns the number of children. Returns 0 if the instance represents
+   * an object.
+   */
+  std::size_t GetChildrenCount() const {
+    if (IsFolder()) return children.size();
+    return 0;
+  }
+  /**
+   * \brief Returns the child ObjectFolderOrObject at the given index.
+   */
+  ObjectFolderOrObject& GetChildAt(std::size_t index);
+  /**
+   * \brief Returns the child ObjectFolderOrObject that represents the object
+   * with the given name. To use only if sure that the instance holds the object
+   * in its direct children (no recursive search).
+   *
+   * \note The equivalent method to get a folder by its name cannot be
+   * implemented because there is no unicity enforced on the folder name.
+   */
+  ObjectFolderOrObject& GetObjectChild(const gd::String& name);
+
+  /**
+   * \brief Returns the parent of the instance. If the instance has no parent
+   * (root folder), the null object is returned.
+   */
+  ObjectFolderOrObject& GetParent() {
+    if (parent == nullptr) {
+      return badObjectFolderOrObject;
+    }
+    return *parent;
+  };
+
+  /**
+   * \brief Moves a child from a position to a new one.
+   */
+  void MoveChild(std::size_t oldIndex, std::size_t newIndex);
+  /**
+   * \brief Removes the given child from the instance's children. If the given
+   * child contains children of its own, does nothing.
+   */
+  void RemoveFolderChild(gd::ObjectFolderOrObject& childToRemove);
+  /**
+   * \brief Removes the child representing the object with the given name from
+   * the instance children and recursively does it for every folder children.
+   */
+  void RemoveRecursivelyObjectNamed(const gd::String& name);
+
+  /**
+   * \brief Inserts an instance representing the given object at the end of the
+   * children.
+   */
+  void InsertObject(gd::Object* insertedObject);
+  /**
+   * \brief Inserts an instance representing the given object at the given
+   * position.
+   */
+  void InsertObject(gd::Object* insertedObject, std::size_t position);
```

Comment:
```suggestion
  /**
   * \brief Inserts an instance representing the given object at the given
   * position.
   */
  void InsertObject(gd::Object* insertedObject, std::size_t position = (size_t)-1);
```

which means that if you don't pass position, C++ will replace it with the maximum position (because size_t is unsigned, so -1 is the max value), which will insert it at the end (you have to adapt the method to handle this)